File: apps/api/src/routes/owner/settlements.ts
// @ts-nocheck
import type { FastifyPluginAsync } from 'fastify';
import type { ZodTypeProvider } from 'fastify-type-provider-zod';
import { z } from 'zod';

export default (async function ownerSettlementRoutes(fastify, opts) {
  const { db, messageBus } = opts as any;

  // Add auth hook
  fastify.addHook('onRequest', async (request, reply) => {
    try {
      await request.jwtVerify();
      const user = request.user as any;
      if (user.role !== 'owner') {
        return reply.status(403).send({ error: 'Owner only' });
      }
      
      const { locationId } = request.params as any;
      if (!locationId || !user.activeLocationId || user.activeLocationId !== locationId) {
        return reply.status(404).send({ error: 'Not found' }); // Tenant isolation
      }
    } catch (err) {
      return reply.status(401).send({ error: 'Unauthorized' });
    }
  });

  // GET /api/owner/locations/:locationId/settlements
  fastify.get('/:locationId/settlements', {
    schema: {
      params: z.object({ locationId: z.string().uuid() }),
      querystring: z.object({
        status: z.enum(['pending', 'approved', 'paid', 'disputed']).optional(),
        courier_id: z.string().uuid().optional(),
        period_start: z.string().optional(),
        period_end: z.string().optional()
      })
    }
  }, async (request, reply) => {
    const { locationId } = request.params;
    const q = request.query;
    
    let sql = `
      SELECT p.id, p.courier_id, c.full_name_encrypted, p.deliveries_count, p.total_earned,
             p.period_start, p.period_end, p.status, p.created_at, p.approved_at, p.paid_at,
             (SELECT currency_code FROM locations WHERE id = $1) as currency
      FROM courier_payouts p
      JOIN couriers c ON c.id = p.courier_id
      WHERE p.location_id = $1
    `;
    const params: any[] = [locationId];
    let idx = 2;

    if (q.status) { sql += ` AND p.status = $${idx++}`; params.push(q.status); }
    if (q.courier_id) { sql += ` AND p.courier_id = $${idx++}`; params.push(q.courier_id); }
    if (q.period_start) { sql += ` AND p.period_start >= $${idx++}`; params.push(q.period_start); }
    if (q.period_end) { sql += ` AND p.period_end <= $${idx++}`; params.push(q.period_end); }

    sql += ` ORDER BY p.created_at DESC`;

    const { decryptPII } = await import('../../lib/pii-cipher.js');

    const res = await db.query(sql, params);
    const payouts = res.rows.map(r => {
      const name = decryptPII(r.full_name_encrypted) || '';
      return {
        id: r.id,
        courierId: r.courier_id,
        courierNameMasked: name ? name.charAt(0) + '***' : 'A***',
        deliveriesCount: r.deliveries_count,
        totalEarned: r.total_earned,
        currency: r.currency,
        periodStart: r.period_start,
        periodEnd: r.period_end,
        status: r.status,
        createdAt: r.created_at,
        approvedAt: r.approved_at,
        paidAt: r.paid_at
      };
    });

    return { payouts };
  });

  // GET /api/owner/locations/:locationId/settlements/:id
  fastify.get('/:locationId/settlements/:id', {
    schema: {
      params: z.object({ locationId: z.string().uuid(), id: z.string().uuid() })
    }
  }, async (request, reply) => {
    const { locationId, id } = request.params;
    
    const payoutRes = await db.query(`
      SELECT p.*, c.full_name_encrypted
      FROM courier_payouts p
      JOIN couriers c ON c.id = p.courier_id
      WHERE p.id = $1 AND p.location_id = $2
    `, [id, locationId]);

    if (payoutRes.rowCount === 0) return reply.status(404).send({ error: 'Not found' });

    const itemsRes = await db.query(`
      SELECT si.assignment_id, ca.order_id, ca.delivered_at, si.amount, si.currency_code
      FROM settlement_items si
      JOIN courier_assignments ca ON ca.id = si.assignment_id
      WHERE si.payout_id = $1
      ORDER BY ca.delivered_at DESC
    `, [id]);

    const items = itemsRes.rows.map(r => ({
      shortOrderId: r.order_id.substring(0, 8),
      deliveredAt: r.delivered_at,
      amount: r.amount,
      currency: r.currency_code,
    }));

    return { payout: payoutRes.rows[0], items };
  });

  // POST approve
  fastify.post('/:locationId/settlements/:id/approve', {
    schema: { params: z.object({ locationId: z.string().uuid(), id: z.string().uuid() }) },
    config: { rateLimit: { max: 30, timeWindow: '1 minute' } }
  }, async (request, reply) => {
    const { locationId, id } = request.params;
    const userId = (request.user as any).sub;

    const client = await db.connect();
    try {
      await client.query('BEGIN');

      const pRes = await client.query(`
        UPDATE courier_payouts
        SET status = 'approved', approved_at = now()
        WHERE id = $1 AND location_id = $2 AND status = 'pending'
        RETURNING *
      `, [id, locationId]);

      if (pRes.rowCount === 0) {
        await client.query('ROLLBACK');
        return reply.status(409).send({ error: 'Payout not found or not pending' });
      }

      await client.query(`
        INSERT INTO settlement_audit_log (payout_id, location_id, action, actor_kind, actor_id)
        VALUES ($1, $2, 'approved', 'owner', $3)
      `, [id, locationId, userId]);

      await client.query('COMMIT');

      // Publish event
      const p = pRes.rows[0];
      await messageBus.publish('settlement.approved', {
        payoutId: p.id,
        courierId: p.courier_id,
        locationId: p.location_id,
        totalEarned: p.total_earned,
        currency: 'ALL', // In variant single currency
        periodStart: p.period_start,
        periodEnd: p.period_end
      });

      return reply.status(200).send({ success: true });
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  });

  // POST pay
  fastify.post('/:locationId/settlements/:id/pay', {
    schema: { 
      params: z.object({ locationId: z.string().uuid(), id: z.string().uuid() }),
      body: z.object({ payment_reference: z.string().optional(), payment_method: z.enum(['cash', 'bank_transfer', 'other']).optional() })
    },
    config: { rateLimit: { max: 30, timeWindow: '1 minute' } }
  }, async (request, reply) => {
    const { locationId, id } = request.params;
    const { payment_reference, payment_method } = request.body;
    const userId = (request.user as any).sub;

    const client = await db.connect();
    try {
      await client.query('BEGIN');

      const pRes = await client.query(`
        UPDATE courier_payouts
        SET status = 'paid', paid_at = now()
        WHERE id = $1 AND location_id = $2 AND status = 'approved'
        RETURNING *
      `, [id, locationId]);

      if (pRes.rowCount === 0) {
        await client.query('ROLLBACK');
        return reply.status(409).send({ error: 'Payout not found or not approved' });
      }

      await client.query(`
        INSERT INTO settlement_audit_log (payout_id, location_id, action, actor_kind, actor_id, metadata)
        VALUES ($1, $2, 'paid', 'owner', $3, $4)
      `, [id, locationId, userId, JSON.stringify({ payment_reference, payment_method })]);

      await client.query('COMMIT');

      return reply.status(200).send({ success: true });
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  });

  // POST dispute
  fastify.post('/:locationId/settlements/:id/dispute', {
    schema: { 
      params: z.object({ locationId: z.string().uuid(), id: z.string().uuid() }),
      body: z.object({ reason: z.string(), items: z.array(z.string().uuid()).optional() })
    },
    config: { rateLimit: { max: 10, timeWindow: '1 minute' } }
  }, async (request, reply) => {
    const { locationId, id } = request.params;
    const { reason, items } = request.body;
    const userId = (request.user as any).sub;

    const client = await db.connect();
    try {
      await client.query('BEGIN');

      const pRes = await client.query(`
        UPDATE courier_payouts
        SET status = 'disputed'
        WHERE id = $1 AND location_id = $2 AND status IN ('pending', 'approved')
        RETURNING *
      `, [id, locationId]);

      if (pRes.rowCount === 0) {
        await client.query('ROLLBACK');
        return reply.status(409).send({ error: 'ALREADY_DISPUTED or invalid status' });
      }

      await client.query(`
        INSERT INTO settlement_audit_log (payout_id, location_id, action, actor_kind, actor_id, metadata)
        VALUES ($1, $2, 'disputed', 'owner', $3, $4)
      `, [id, locationId, userId, JSON.stringify({ reason, disputed_items: items })]);

      await client.query('COMMIT');

      // Notify courier (Telegram) handled elsewhere by listening to a dispute event, or just emit event
      await messageBus.publish('settlement.disputed', {
        payoutId: id,
        courierId: pRes.rows[0].courier_id,
        locationId
      });

      return reply.status(200).send({ success: true });
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  });

  // POST reopen
  fastify.post('/:locationId/settlements/:id/reopen', {
    schema: { 
      params: z.object({ locationId: z.string().uuid(), id: z.string().uuid() }),
      body: z.object({ reason: z.string() })
    },
    config: { rateLimit: { max: 10, timeWindow: '1 minute' } }
  }, async (request, reply) => {
    const { locationId, id } = request.params;
    const { reason } = request.body;
    const userId = (request.user as any).sub;

    const client = await db.connect();
    try {
      await client.query('BEGIN');

      const pRes = await client.query(`
        UPDATE courier_payouts
        SET status = 'pending'
        WHERE id = $1 AND location_id = $2 AND status = 'disputed'
        RETURNING *
      `, [id, locationId]);

      if (pRes.rowCount === 0) {
        await client.query('ROLLBACK');
        return reply.status(409).send({ error: 'Payout not disputed' });
      }

      await client.query(`
        INSERT INTO settlement_audit_log (payout_id, location_id, action, actor_kind, actor_id, metadata)
        VALUES ($1, $2, 'reopened', 'owner', $3, $4)
      `, [id, locationId, userId, JSON.stringify({ reason })]);

      await client.query('COMMIT');

      return reply.status(200).send({ success: true });
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  });

  // POST regenerate manually
  fastify.post('/:locationId/settlements/regenerate', {
    schema: { 
      params: z.object({ locationId: z.string().uuid() }),
      body: z.object({ referenceDate: z.string() })
    },
    config: { rateLimit: { max: 5, timeWindow: '5 minutes' } }
  }, async (request, reply) => {
    const { locationId } = request.params;
    const { referenceDate } = request.body;
    
    // Publish a job specifically for this location or globally
    const { SettlementCronWorker } = await import('../../workers/settlement-cron.js');
    const worker = new SettlementCronWorker(db, null as any);
    await worker.handleGenerate(new Date(referenceDate)); // Technically processes all locations. For scale, we'd limit to locationId. 

    return reply.status(200).send({ success: true });
  });

}) as FastifyPluginAsync<any, any, ZodTypeProvider>;

File: rebuild/crates/api/src/routes/courier/settlements.rs
//! S7 courier settlements — the courier's OWN read of their `courier_payouts`/`settlement_items`
//! rows. Ports `apps/api/src/routes/courier/settlements.ts`. See `crate::routes::courier` module
//! doc for the shared auth/tenancy contract (`CourierSession` extractor +
//! `with_tenant(active_location_id)`), and `courier::me`/`courier::assignments` for sibling S7
//! files using the same conventions.
//!
//! ## Money is NOT generated here (REV-S7-3, S7-T6)
//! Settlement GENERATION (`app_generate_settlements`, S5/S8-owned) is never called or
//! re-implemented by this file — every method here is a pure READ of rows some OTHER surface
//! already wrote. No settlement math, no `total_earned`/`amount` computation, anywhere below.
//!
//! ## Tenancy fix-by-port (REV-S7-1 / S7-T10)
//! Old Node's `settlements.ts` seats `app.current_tenant` on the BARE pool with NO `BEGIN`
//! (`SELECT set_config('app.current_tenant', $1, true)` as a standalone auto-committed statement,
//! `settlements.ts:25,59,74`) — the seat can land on a different pooled connection than the
//! subsequent SELECT, so it only "works" today because the operational pool's role still bypasses
//! RLS. `settlement_items`'s RLS policy is additionally the THROWING bare `current_setting(...)`
//! form (no `, true` missing-ok arg, `settlement-items.ts` migration) — under real NOBYPASSRLS
//! enforcement a missing seat would 500, not silently return zero rows. Every method below runs
//! inside [`crate::db::with_tenant`] (a real `BEGIN -> set_config -> ... -> COMMIT` transaction),
//! closing both gaps by construction.
//!
//! ## Q-PAYOUT-READ-SHARED — stricter courier redaction on payout items
//! `get_payout_detail`'s items list is a SEPARATE, narrower query than what an owner-facing
//! settlement read would return: "strictly no orderId, no assignmentId, no customer phone"
//! (`settlements.ts:89`, carried verbatim as this file's own comment below). [`PayoutItemRow`]'s
//! fields are `delivered_at`/`amount`/`currency` ONLY — `payout_items_never_include_order_id_or_
//! assignment_id` is the structural DoD test pinning this shape.
//!
//! ## Wire shape: RAW snake_case passthrough, no camelCase mapper (unlike `courier::me`)
//! Node's `settlements.ts` sends `res.rows` straight through with no `mapXRow` transform at all —
//! [`PayoutRow`]/[`PayoutItemRow`] double as BOTH the repo row type and the wire DTO (their field
//! names literally ARE the SQL column aliases), matching that absence of a mapping layer exactly.

use std::sync::Arc;

use axum::Json;
use axum::extract::{Extension, Path, Query};
use axum::response::IntoResponse;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tower_http::request_id::RequestId;
use uuid::Uuid;

use domain::{ErrorCode, TenantId};

use crate::auth::AuthState;
use crate::auth::extractors::CourierSession;
use crate::error::ApiError;
use crate::repo::RepoError;
use crate::routes::correlation_id_string;

// ── State + repo trait ──────────────────────────────────────────────────────────────────────

#[derive(Clone)]
pub struct SettlementsState {
    /// Not read here — payout handlers authenticate via the `CourierSession` extractor (which reads
    /// `AuthState` from request extensions, layered by `courier_router`). Kept for State-shape
    /// uniformity with the other courier submodules; see `assignments::AssignmentsState.auth`.
    #[allow(
        dead_code,
        reason = "CourierSession extractor reads AuthState from request extensions, not this field — kept for State-shape uniformity"
    )]
    pub auth: AuthState,
    pub repo: Arc<dyn SettlementsRepo>,
}

/// `courier_payouts JOIN locations` (`settlements.ts:28-34`, `:62-68`) — RAW column-aliased shape,
/// no camelCase transform; doubles as the wire DTO (see module doc).
#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
pub struct PayoutRow {
    pub id: Uuid,
    pub location_id: Uuid,
    pub location_name: String,
    pub deliveries_count: i32,
    pub total_earned: i64,
    #[serde(serialize_with = "crate::dto::serialize_js_instant")]
    pub period_start: DateTime<Utc>,
    #[serde(serialize_with = "crate::dto::serialize_js_instant")]
    pub period_end: DateTime<Utc>,
    pub status: String,
    #[serde(serialize_with = "crate::dto::serialize_js_instant")]
    pub created_at: DateTime<Utc>,
    #[serde(serialize_with = "crate::dto::serialize_js_instant_opt")]
    pub approved_at: Option<DateTime<Utc>>,
    pub currency: String,
}

/// `settlement_items JOIN courier_assignments` (`settlements.ts:79-85`) — Q-PAYOUT-READ-SHARED:
/// strictly no orderId, no assignmentId, no customer phone. These three fields are the WHOLE
/// shape by design; the structural test pins that no more ever get added silently.
#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
pub struct PayoutItemRow {
    #[serde(serialize_with = "crate::dto::serialize_js_instant_opt")]
    pub delivered_at: Option<DateTime<Utc>>,
    pub amount: i64,
    pub currency: String,
}

#[async_trait::async_trait]
pub trait SettlementsRepo: Send + Sync {
    /// `GET /me/payouts` (`settlements.ts:12-48`) — optional `status` filter.
    async fn list_payouts(
        &self,
        courier_id: Uuid,
        location_id: Uuid,
        status: Option<String>,
    ) -> Result<Vec<PayoutRow>, RepoError>;

    /// `GET /me/payouts/:id`'s payout half (`settlements.ts:62-71`) — scoped to
    /// `(id, courier_id)`. `Ok(None)` covers BOTH "no such payout" AND "belongs to another
    /// courier" (the actor-gate's existence-hiding posture, same reasoning as S3/S7's other
    /// not-found collapses) — 404 either way.
    async fn get_payout(
        &self,
        id: Uuid,
        courier_id: Uuid,
        location_id: Uuid,
    ) -> Result<Option<PayoutRow>, RepoError>;

    /// `GET /me/payouts/:id`'s items half (`settlements.ts:79-85`).
    async fn payout_items(
        &self,
        payout_id: Uuid,
        location_id: Uuid,
    ) -> Result<Vec<PayoutItemRow>, RepoError>;
}

// ── DTOs ─────────────────────────────────────────────────────────────────────────────────────

/// `?status=` querystring (`settlements.ts:14-16`).
#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum PayoutStatus {
    Pending,
    Approved,
    Paid,
    Disputed,
}

impl PayoutStatus {
    fn as_db_str(self) -> &'static str {
        match self {
            PayoutStatus::Pending => "pending",
            PayoutStatus::Approved => "approved",
            PayoutStatus::Paid => "paid",
            PayoutStatus::Disputed => "disputed",
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct ListPayoutsQuery {
    #[serde(default)]
    pub status: Option<PayoutStatus>,
}

// `Serialize`-only (no `ToSchema`): `PayoutRow`/`PayoutItemRow` carry raw `DateTime<Utc>` fields,
// which don't implement utoipa's `PartialSchema` without a workspace-wide Cargo.toml "chrono"
// feature this two-file port doesn't add (see `courier::me`'s identical note, which cites the
// established `owner::products::ProductRow` precedent for this same workaround).
#[derive(Debug, Serialize)]
pub struct PayoutsListResponse {
    pub payouts: Vec<PayoutRow>,
}

#[derive(Debug, Serialize)]
pub struct PayoutDetailResponse {
    pub payout: PayoutRow,
    pub items: Vec<PayoutItemRow>,
}

fn internal_error(correlation_id: String) -> ApiError {
    ApiError::new(ErrorCode::Internal, "internal_error", correlation_id)
}

fn not_found(correlation_id: String) -> ApiError {
    ApiError::new(ErrorCode::NotFound, "Not found", correlation_id)
}

// ── Handlers ─────────────────────────────────────────────────────────────────────────────────

/// `GET /api/courier/me/payouts` (`settlements.ts:12-48`).
#[utoipa::path(get, path = "/api/courier/me/payouts", tag = "courier",
    params(("status" = Option<String>, Query, description = "pending|approved|paid|disputed")),
    responses((status = 200, description = "This courier's settlement payouts")))]
pub async fn get_payouts(
    Extension(state): Extension<SettlementsState>,
    CourierSession(courier): CourierSession,
    Query(params): Query<ListPayoutsQuery>,
    Extension(request_id): Extension<RequestId>,
) -> Result<impl IntoResponse, ApiError> {
    let correlation_id = correlation_id_string(&request_id);
    let status = params.status.map(|s| s.as_db_str().to_string());
    let rows = state
        .repo
        .list_payouts(courier.sub, courier.active_location_id, status)
        .await
        .map_err(|_e| internal_error(correlation_id))?;
    Ok(Json(PayoutsListResponse { payouts: rows }))
}

/// `GET /api/courier/me/payouts/{id}` (`settlements.ts:51-91`) — Q-PAYOUT-READ-SHARED redaction
/// on `items`.
#[utoipa::path(get, path = "/api/courier/me/payouts/{id}", tag = "courier",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, description = "Payout + redacted settlement items"),
        (status = 404, description = "Not found or belongs to another courier", body = domain::ErrorEnvelope),
    ))]
pub async fn get_payout_detail(
    Extension(state): Extension<SettlementsState>,
    CourierSession(courier): CourierSession,
    Path(id): Path<Uuid>,
    Extension(request_id): Extension<RequestId>,
) -> Result<impl IntoResponse, ApiError> {
    let correlation_id = correlation_id_string(&request_id);
    let payout = state
        .repo
        .get_payout(id, courier.sub, courier.active_location_id)
        .await
        .map_err(|_e| internal_error(correlation_id.clone()))?
        .ok_or_else(|| not_found(correlation_id.clone()))?;

    let items = state
        .repo
        .payout_items(id, courier.active_location_id)
        .await
        .map_err(|_e| internal_error(correlation_id))?;

    Ok(Json(PayoutDetailResponse { payout, items }))
}

// ── PgSettlementsRepo ────────────────────────────────────────────────────────────────────────

const PAYOUT_SELECT: &str = "SELECT p.id, p.location_id, l.name AS location_name, p.deliveries_count, \
     p.total_earned::bigint AS total_earned, p.period_start, p.period_end, p.status, p.created_at, \
     p.approved_at, l.currency_code AS currency \
     FROM courier_payouts p JOIN locations l ON l.id = p.location_id ";

pub struct PgSettlementsRepo {
    pool: sqlx::PgPool,
}

impl PgSettlementsRepo {
    pub fn new(pool: sqlx::PgPool) -> Self {
        PgSettlementsRepo { pool }
    }
}

fn map_txn_err(err: crate::db::TenantTxnError) -> RepoError {
    use crate::db::TenantTxnError;
    match err {
        TenantTxnError::Begin(e)
        | TenantTxnError::SetTenant(e)
        | TenantTxnError::Work(e)
        | TenantTxnError::Commit(e) => RepoError(e),
        TenantTxnError::WorkThenRollbackFailed { work, .. } => RepoError(work),
    }
}

#[async_trait::async_trait]
impl SettlementsRepo for PgSettlementsRepo {
    async fn list_payouts(
        &self,
        courier_id: Uuid,
        location_id: Uuid,
        status: Option<String>,
    ) -> Result<Vec<PayoutRow>, RepoError> {
        crate::db::with_tenant(&self.pool, TenantId::from(location_id), move |txn| {
            Box::pin(async move {
                let rows: Vec<PayoutRow> = if let Some(status) = status {
                    let sql = format!("{PAYOUT_SELECT}WHERE p.courier_id = $1 AND p.status = $2 ORDER BY p.created_at DESC");
                    sqlx::query_as(&sql)
                        .bind(courier_id)
                        .bind(status)
                        .fetch_all(&mut **txn)
                        .await?
                } else {
                    let sql = format!("{PAYOUT_SELECT}WHERE p.courier_id = $1 ORDER BY p.created_at DESC");
                    sqlx::query_as(&sql)
                        .bind(courier_id)
                        .fetch_all(&mut **txn)
                        .await?
                };
                Ok(rows)
            })
        })
        .await
        .map_err(map_txn_err)
    }

    async fn get_payout(
        &self,
        id: Uuid,
        courier_id: Uuid,
        location_id: Uuid,
    ) -> Result<Option<PayoutRow>, RepoError> {
        crate::db::with_tenant(&self.pool, TenantId::from(location_id), move |txn| {
            Box::pin(async move {
                let sql = format!("{PAYOUT_SELECT}WHERE p.id = $1 AND p.courier_id = $2");
                let row: Option<PayoutRow> = sqlx::query_as(&sql)
                    .bind(id)
                    .bind(courier_id)
                    .fetch_optional(&mut **txn)
                    .await?;
                Ok(row)
            })
        })
        .await
        .map_err(map_txn_err)
    }

    async fn payout_items(
        &self,
        payout_id: Uuid,
        location_id: Uuid,
    ) -> Result<Vec<PayoutItemRow>, RepoError> {
        crate::db::with_tenant(&self.pool, TenantId::from(location_id), move |txn| {
            Box::pin(async move {
                // Strictly no orderId, no assignmentId, no customer phone (settlements.ts:89) —
                // the SELECT list itself is the redaction, not a post-hoc field strip.
                let rows: Vec<PayoutItemRow> = sqlx::query_as(
                    "SELECT ca.delivered_at, si.amount::bigint AS amount, si.currency_code AS currency \
                     FROM settlement_items si JOIN courier_assignments ca ON ca.id = si.assignment_id \
                     WHERE si.payout_id = $1 ORDER BY ca.delivered_at DESC",
                )
                .bind(payout_id)
                .fetch_all(&mut **txn)
                .await?;
                Ok(rows)
            })
        })
        .await
        .map_err(map_txn_err)
    }
}

// ── FakeSettlementsRepo (test-only) ──────────────────────────────────────────────────────────

#[cfg(test)]
pub mod fake {
    use super::{PayoutItemRow, PayoutRow, RepoError, SettlementsRepo};
    use std::collections::HashMap;
    use std::sync::Mutex;
    use uuid::Uuid;

    #[derive(Default)]
    pub struct FakeSettlementsRepo {
        /// payout_id -> (owning courier_id, row). The courier scoping lives here (not on
        /// `PayoutRow`, which never carries a `courier_id` column — matching Node's own SELECT).
        pub payouts: Mutex<HashMap<Uuid, (Uuid, PayoutRow)>>,
        pub items: Mutex<HashMap<Uuid, Vec<PayoutItemRow>>>,
    }

    impl FakeSettlementsRepo {
        pub fn seed_payout(&self, courier_id: Uuid, row: PayoutRow) {
            self.payouts
                .lock()
                .unwrap()
                .insert(row.id, (courier_id, row));
        }
        pub fn seed_items(&self, payout_id: Uuid, items: Vec<PayoutItemRow>) {
            self.items.lock().unwrap().insert(payout_id, items);
        }
    }

    #[async_trait::async_trait]
    impl SettlementsRepo for FakeSettlementsRepo {
        async fn list_payouts(
            &self,
            courier_id: Uuid,
            _location_id: Uuid,
            status: Option<String>,
        ) -> Result<Vec<PayoutRow>, RepoError> {
            Ok(self
                .payouts
                .lock()
                .unwrap()
                .values()
                .filter(|(cid, row)| {
                    *cid == courier_id && status.as_deref().is_none_or(|s| row.status == s)
                })
                .map(|(_, row)| row.clone())
                .collect())
        }

        async fn get_payout(
            &self,
            id: Uuid,
            courier_id: Uuid,
            _location_id: Uuid,
        ) -> Result<Option<PayoutRow>, RepoError> {
            Ok(self
                .payouts
                .lock()
                .unwrap()
                .get(&id)
                .filter(|(cid, _)| *cid == courier_id)
                .map(|(_, row)| row.clone()))
        }

        async fn payout_items(
            &self,
            payout_id: Uuid,
            _location_id: Uuid,
        ) -> Result<Vec<PayoutItemRow>, RepoError> {
            Ok(self
                .items
                .lock()
                .unwrap()
                .get(&payout_id)
                .cloned()
                .unwrap_or_default())
        }
    }
}

// ── Tests ────────────────────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::fake::FakeSettlementsRepo;
    use super::*;
    use crate::auth::claims::CourierClaims;
    use crate::auth::repo::fake::FakeAuthRepo;
    use axum::body::to_bytes;
    use axum::http::StatusCode;
    use axum::response::Response;
    use std::sync::Arc;

    fn request_id() -> RequestId {
        RequestId::new(axum::http::HeaderValue::from_static("corr-1"))
    }

    fn courier_session(courier_id: Uuid, location_id: Uuid) -> CourierSession {
        CourierSession(CourierClaims::new(courier_id, location_id, None))
    }

    fn state_with(repo: FakeSettlementsRepo) -> SettlementsState {
        SettlementsState {
            auth: AuthState::test_state(Arc::new(FakeAuthRepo::default())),
            repo: Arc::new(repo),
        }
    }

    async fn json_body(resp: Response) -> (StatusCode, serde_json::Value) {
        let status = resp.status();
        let bytes = to_bytes(resp.into_body(), usize::MAX).await.unwrap();
        (status, serde_json::from_slice(&bytes).unwrap())
    }

    fn fixture_payout(id: Uuid, location_id: Uuid, status: &str) -> PayoutRow {
        PayoutRow {
            id,
            location_id,
            location_name: "Eljo's Pizza".to_string(),
            deliveries_count: 12,
            total_earned: 4500,
            period_start: Utc::now(),
            period_end: Utc::now(),
            status: status.to_string(),
            created_at: Utc::now(),
            approved_at: None,
            currency: "ALL".to_string(),
        }
    }

    // ── get_payouts ──

    #[tokio::test]
    async fn get_payouts_200_happy_path() {
        let courier_id = Uuid::new_v4();
        let location = Uuid::new_v4();
        let repo = FakeSettlementsRepo::default();
        repo.seed_payout(
            courier_id,
            fixture_payout(Uuid::new_v4(), location, "pending"),
        );
        let state = state_with(repo);

        let resp = get_payouts(
            Extension(state),
            courier_session(courier_id, location),
            Query(ListPayoutsQuery { status: None }),
            Extension(request_id()),
        )
        .await
        .unwrap()
        .into_response();
        let (status, json) = json_body(resp).await;
        assert_eq!(status, StatusCode::OK);
        assert_eq!(json["payouts"].as_array().unwrap().len(), 1);
    }

    #[tokio::test]
    async fn get_payouts_filters_by_status_query_param() {
        let courier_id = Uuid::new_v4();
        let location = Uuid::new_v4();
        let repo = FakeSettlementsRepo::default();
        repo.seed_payout(
            courier_id,
            fixture_payout(Uuid::new_v4(), location, "pending"),
        );
        repo.seed_payout(courier_id, fixture_payout(Uuid::new_v4(), location, "paid"));
        // A payout belonging to ANOTHER courier must never leak into this courier's filtered list.
        repo.seed_payout(
            Uuid::new_v4(),
            fixture_payout(Uuid::new_v4(), location, "paid"),
        );
        let state = state_with(repo);

        let resp = get_payouts(
            Extension(state),
            courier_session(courier_id, location),
            Query(ListPayoutsQuery {
                status: Some(PayoutStatus::Paid),
            }),
            Extension(request_id()),
        )
        .await
        .unwrap()
        .into_response();
        let (_, json) = json_body(resp).await;
        let payouts = json["payouts"].as_array().unwrap();
        assert_eq!(payouts.len(), 1);
        assert_eq!(payouts[0]["status"], "paid");
    }

    // ── get_payout_detail ──

    #[tokio::test]
    async fn get_payout_detail_200_happy_path_returns_payout_and_items() {
        let courier_id = Uuid::new_v4();
        let location = Uuid::new_v4();
        let payout_id = Uuid::new_v4();
        let repo = FakeSettlementsRepo::default();
        repo.seed_payout(courier_id, fixture_payout(payout_id, location, "paid"));
        repo.seed_items(
            payout_id,
            vec![PayoutItemRow {
                delivered_at: Some(Utc::now()),
                amount: 900,
                currency: "ALL".to_string(),
            }],
        );
        let state = state_with(repo);

        let resp = get_payout_detail(
            Extension(state),
            courier_session(courier_id, location),
            Path(payout_id),
            Extension(request_id()),
        )
        .await
        .unwrap()
        .into_response();
        let (status, json) = json_body(resp).await;
        assert_eq!(status, StatusCode::OK);
        assert_eq!(json["payout"]["id"], payout_id.to_string());
        assert_eq!(json["items"].as_array().unwrap().len(), 1);
        assert_eq!(json["items"][0]["amount"], 900);
    }

    #[tokio::test]
    async fn get_payout_detail_404_for_a_foreign_payout() {
        let owner_courier = Uuid::new_v4();
        let attacker_courier = Uuid::new_v4();
        let location = Uuid::new_v4();
        let payout_id = Uuid::new_v4();
        let repo = FakeSettlementsRepo::default();
        repo.seed_payout(owner_courier, fixture_payout(payout_id, location, "paid"));
        let state = state_with(repo);

        let err = crate::error::expect_err(
            get_payout_detail(
                Extension(state),
                courier_session(attacker_courier, location),
                Path(payout_id),
                Extension(request_id()),
            )
            .await,
        );
        assert_eq!(err.envelope.code, ErrorCode::NotFound);
    }

    /// Structural DoD test (Q-PAYOUT-READ-SHARED): the payout-items DTO must literally have no
    /// `orderId`/`assignmentId`/customer-phone field — pin the serialized key set exactly.
    #[test]
    fn get_payout_detail_items_never_include_order_id_or_assignment_id() {
        let item = PayoutItemRow {
            delivered_at: Some(Utc::now()),
            amount: 500,
            currency: "ALL".to_string(),
        };
        let value = serde_json::to_value(&item).unwrap();
        let keys: std::collections::BTreeSet<String> =
            value.as_object().unwrap().keys().cloned().collect();
        let expected: std::collections::BTreeSet<String> = ["delivered_at", "amount", "currency"]
            .into_iter()
            .map(str::to_string)
            .collect();
        assert_eq!(
            keys, expected,
            "payout items must strictly exclude orderId/assignmentId/customer phone"
        );
    }
}
